Add an API function to get the UTF-8 encoded string size.

robertsipka opened the pull request.

zherczeg commented on jerry-core/ecma/base/ecma-helpers-string.c:
```diff
@@ -1380,6 +1380,42 @@ ecma_string_get_size (const ecma_string_t *string_p) /**< ecma-string */
+    default:
+    {
+      JERRY_ASSERT (ECMA_STRING_GET_CONTAINER (string_p) == ECMA_STRING_CONTAINER_MAGIC_STRING_EX);
+      return lit_get_magic_string_ex_size (string_p->u.magic_string_ex_id);
```
Actually there is no guarantee that the external magic strings does not contain surrogate pairs.

zherczeg commented on jerry-core/lit/lit-strings.c:
```diff
@@ -282,6 +282,34 @@ lit_utf8_string_length (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string */
+
+  while (position < utf8_buf_size)
+  {
+    lit_utf8_size_t char_size = lit_get_unicode_char_size_by_utf8_first_byte (*(utf8_buf_p + position));
```
utf8_buf_p[position]

I would prefer offset rather than position.

robertsipka commented:
@zherczeg : Thanks for the review, I accidentally ignored the external magic strings. I've updated the patch.

LaszloLango commented:
LGTM

zherczeg commented on jerry-core/ecma/base/ecma-helpers-string.c:
```diff
@@ -1380,6 +1380,44 @@ ecma_string_get_size (const ecma_string_t *string_p) /**< ecma-string */
+    default:
+    {
+      JERRY_ASSERT (ECMA_STRING_GET_CONTAINER (string_p) == ECMA_STRING_CONTAINER_MAGIC_STRING_EX);
+      return lit_calculate_utf8_string_size_from_cesu8_string (
```
Please follow the style.

robertsipka commented:
@dbatyai : Indeed, thanks! I've added this check.
@zherczeg, @LaszloLango : I've rebased the patch and modified the calculation part considering the surrogate pairs, and also extended the api tests with a string test which contains a three-byte long sequence, where the char isn't  part of a surrogate pair.

zherczeg commented on jerry-core/ecma/base/ecma-helpers-string.c:
```diff
@@ -1514,6 +1514,62 @@ ecma_string_get_size (const ecma_string_t *string_p) /**< ecma-string */
+    case ECMA_STRING_CONTAINER_HEAP_UTF8_STRING:
+    {
+      if (ECMA_STRING_IS_ASCII ((const lit_utf8_byte_t *) (string_p + 1),
+                                (lit_utf8_size_t) string_p->u.utf8_string.size))
```
Perhaps we could simply compare (lit_utf8_size_t) string_p->u.utf8_string.size with (lit_utf8_size_t) string_p->u.utf8_string.length. I feel calling a function is unnecessary because the check is very simple.

zherczeg commented on jerry-core/ecma/base/ecma-helpers-string.c:
```diff
@@ -1514,6 +1514,62 @@ ecma_string_get_size (const ecma_string_t *string_p) /**< ecma-string */
+    }
+    case ECMA_STRING_CONTAINER_HEAP_LONG_UTF8_STRING:
+    {
+      if (ECMA_STRING_IS_ASCII ((const lit_utf8_byte_t *) (string_p + 1),
```
Ditto.

zherczeg commented on jerry-core/lit/lit-strings.c:
```diff
@@ -282,6 +282,43 @@ lit_utf8_string_length (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string */
+ */
+lit_utf8_size_t
+lit_get_utf8_size_of_cesu8_string (const lit_utf8_byte_t *utf8_buf_p, /**< cesu-8 string */
+                                   lit_utf8_size_t utf8_buf_size) /**< string size */
```
Should be cesu8_buf_p cesut8_buf_size

zherczeg commented on jerry-core/lit/lit-strings.c:
```diff
@@ -282,6 +282,43 @@ lit_utf8_string_length (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string */
+                                   lit_utf8_size_t utf8_buf_size) /**< string size */
+{
+  lit_utf8_size_t offset = 0;
+  lit_utf8_size_t reduction_size = 0;
```
What about creating a variable here:

    lit_utf8_size_t utf8_buf_size = cesu8_buf_size;

And simply update its value:

    utf8_buf_size -= 2;

zherczeg commented on tests/unit/test-api.c:
```diff
@@ -339,29 +340,62 @@ main (void)
+  TEST_ASSERT (utf8_sz == 13 && cesu8_sz == 17);
+  jerry_release_value (args[0]);
+
+  // Test string: 'str: {DESERET CAPITAL LETTER LONG I}'
```
Only /* */ comments are allowed in Jerry.

robertsipka commented:
@zherczeg : Thanks, I've updated the patch with the suggested modifications.

LaszloLango commented on jerry-core/lit/lit-strings.c:
```diff
@@ -282,6 +282,43 @@ lit_utf8_string_length (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string */
+
+    if (lit_is_code_point_utf16_high_surrogate (ch))
+    {
+      if (offset != cesu8_buf_size)
```
Please merge this 2 if statement

robertsipka replied:
@LaszloLango : done.

zherczeg commented on jerry-core/ecma/base/ecma-helpers-string.c:
```diff
@@ -1514,6 +1514,61 @@ ecma_string_get_size (const ecma_string_t *string_p) /**< ecma-string */
+    {
+      JERRY_ASSERT (ECMA_STRING_GET_CONTAINER (string_p) == ECMA_STRING_CONTAINER_MAGIC_STRING_EX);
+
+      if (ECMA_STRING_IS_ASCII (lit_get_magic_string_ex_utf8 (string_p->u.magic_string_ex_id),
```
I don't think we gain much here since the check is as costly as computing the size. There is no length data for external strings.

zherczeg commented on jerry-core/lit/lit-strings.c:
```diff
@@ -282,6 +282,40 @@ lit_utf8_string_length (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string */
+    ecma_char_t ch;
+    offset += lit_read_code_unit_from_utf8 (cesu8_buf_p + offset, &ch);
+
+    if (lit_is_code_point_utf16_high_surrogate (ch) && (offset != cesu8_buf_size))
```
Use < instead of !=

zherczeg commented on tests/unit/test-api.c:
```diff
@@ -339,29 +340,62 @@ main (void)
-  jerry_size_t cesu8_sz =  jerry_get_string_size (args[1]);
+  /* these size must be equal */
+  utf8_sz = jerry_get_string_size (args[0]);
+  cesu8_sz =  jerry_get_string_size (args[1]);
```
Double spaces

zherczeg commented on tests/unit/test-api.c:
```diff
@@ -339,29 +340,62 @@ main (void)
-  args[0] = jerry_create_string_from_utf8 ((jerry_char_t *) "\x73\x74\x72\x3a\xf0\x9d\x94\xa3\xf0\x9d\x94\xa4");
-  jerry_length_t cesu8_length = jerry_get_string_length (args[0]);
+  /* Test string: 'str: {MATHEMATICAL FRAKTUR SMALL F}{MATHEMATICAL FRAKTUR SMALL G}' */
+  args[0] = jerry_create_string_from_utf8 ((jerry_char_t *) "\x73\x74\x72\x3a \xf0\x9d\x94\xa3\xf0\x9d\x94\xa4");
```
Please add a space before \xf0\x9d\x94\xa4

zherczeg commented:
Quite close to done.

robertsipka commented:
@zherczeg : I've just updated this patch based on your suggestions.

zherczeg commented:
LGTM
